### DISCUSSION TITLE
Where can one publish the wasm files to be used on Ruby on Rails projects?

### DISCUSSION THREAD

**Discussion by benkoshy** [unoplatform/Uno.Wasm.Bootstrap#812] [Q&A] [answered]
The [deploy and publish guide](
https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/main/doc/deploy-and-publish.md) seems to suggest that you need an ASP.NET project.  My situation is different:

1. I have a Ruby on Rails web application.
2. I have a computationally intense method in a .net code base that I control. I want to package this method into a WASM "file" and publish it somewhere so that the JS in my Ruby on rails application can call that wasm method for speedy execution.

- **Question:** How can I do this?
- i.e. Does anyone know where one can securely publish the wasm output so that it can be used in non asp.net projects? 

Any pointers would be much appreciated.

chrs

**Comment by SerratedSharp** [comment, marked as answer]
I could be missing the mark on what you're trying to accomplish, but hopefully this will clear a few things up.

There's not any special server side requirements because the files don't run server side, and are downloaded like regular file downloads, i.e. static file hosting.  There's really no need for a ASP.NET host.  For example I have a WASM bootstrap demo running in Github.io which is nothing but a static file host with no server side code.  See https://serratedsharp.github.io/CSharpWasmJQueryDemo/ and note in network tab of browser console the *.clr WASM files being downloaded.  I think it's similar to your request in that I have a LuhnCheckDigit calculation implemented in .NET WASM that is called from Javascript on the page.

I just took the output of /MyProject/bin/Release/net8.0/dist/ and copied it up to the server.  If Ruby on rails can serve static files like css/images/JS, then it should be able to serve the WASM package files.   You may need to configure mimetypes for "*.clr" as "application/octet-stream" and ".dat" as "application/dat" to allow your web server to serve them. Often web hosts will block downloads of any unknown file extensions until you do whatever is necessary for that web host to add the mimetypes.

To be clear, nothing executes until the WASM package is downloaded to the browser and Uno bootstrap initializes your assembly.  Akin to a javascript file being downloaded and executed in the browser.

All of the output in bin/Release/dist/ (short for "**dist**ribution") are the static files needed, mine might look slightly different as I am using EmebddedMode:
![image](https://github.com/unoplatform/Uno.Wasm.Bootstrap/assets/97156524/079806b8-9bd8-48c3-9694-ba4fe5871c50)

Using the default build, when you access index.html, the browser will load the page, and Uno bootstrap will run and locate the other assets and then execute your WASM Program Main entry point inside the browser. 

So just like you might have a /scripts/ folder on your web server for javascript files, you could create a /wasm/ folder and copy everything from /dist/ into it.  When you navigate to yoururl.com/dist/index.html the WASM will load and execute.

The index.htm is basically a blank page that runs your WASM as if it were a skeleton for a SPA/single page application.  It doesn't integrate in any way with your web application.

So serving the package is the easy part.  Getting the WASM to execute in the same context as whatever other web application code you're wanting to integrate it with is the complicated part.

I would guess what you want to do is from some of your existing client side javascript running inside existing pages of your Ruby on rails app, you want to call a method in your .NET WASM module(which you exported using something like `[JSExport]`).

So maybe you have an computationally intensive CalculateChecksum() function in .NET WASM exported, and whenever someone clicks a button on your ruby rails app, then some javascript button handler runs and you want that javascript to call the CalculateChecksum() function and get the return value.  

I don't think those details matter, but to put in concrete terms, your javascript and the WASM need to run on the same page.

There's several ways to accomplish this, but one way is to use EmbeddedMode switch in your .NET console project that uses Uno.Bootstrap to build to WASM, so adding this to the *.csproj:
```XML
<PropertyGroup>
	<!-- Generate embedded.js -->
	<WasmShellMode>BrowserEmbedded</WasmShellMode>
</PropertyGroup>
```

This changes the output of /dist/ slightly, and includes a embedded.js.

Now from your own application's HTML/JS, you can include this embedded.js assuming you've hosted the dist files on your web server:
```HTML
    <script src="/dist/embedded.js"></script>
    <script src="/js/yourexisting.js"></script>
```

You could adjust the URL to an external host if for some reason it's necessary `<script src="www.someother.com/dist/embedded.js"></script>`. 

When embedded.js executes it will look for or create a <div id='uno-body'></div> and run a loading bar/splash screen while the WASM initializes, then load your assembly and call your Program.cs Main entry point.  Now your [JSExport]'d .NET method should be available for your javascript to call and retrieve whatever it returns.  

https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/features-embedded.mode.html

Note they tell you to declare a div uno-body, but I have always found this results in two duplicate uno body's (mine and an additional one created by embedded.js).

If you are calling the WASM module on demand, such as a triggered by a button click, then great.  If you have some of your own javascript that tries to execute immediately upon page load and wants to call the WASM module immediately, it gets a little more complicated because it takes a moment for initialization and the WASM module might not be "ready" if you try to call it during page load.

**Reply by benkoshy** [reply]
Thank you for your detailed answer.  It has clarified very nicely.  btw cool jquery project. your repo will prove very helpful to me. thank you!

I have a secondary question - if you will permit me. 

I have two separate projects

-  a c# project
- and a rails project.

With normal javascript:

- packages are managed by yarn, or npm etcv
- when I deploy the rails app, these packages are typically transpiled / minified etc.
- and then dumped into a folder where they are served (with a fingerprint)

**Question: ** Do you have any ideas on where i can "publish" all of the output in `bin/Release/dist/ ` so that I can easily pick it up when deploying - i.e. some sort of "pipeline" or a "build pipeline" - much like with a yarn package.  Not sure if I'm making sense - but I wish to avoid manually cutting and pasting the `bin/Release/dist/` output from my windows PC to my rails PC, and then to deploy it.

**Reply by SerratedSharp** [reply]
When you say rails project, is this a Visual Studio project?

I have never gotten this to work **satisfactorily** for me but that may be my own fault, and I don't think it'll work for you cause your destination is not a ASP.NET project, but the documented way to copy on publish is this: https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/deploy-and-publish.html

What I used to do, and I think would work for you, is to have a MSBuild script in your C# "WASM" project that copies on build:
https://github.com/SerratedSharp/SerratedJQ/blob/main/SerratedJQSample/Sample.Wasm/Build.props

In your C# project file you'd add `<Import Project=".\Build.props" />` and this file would be in the C# project.  You'd have to adjust the `DestinationWebProjectName` parameter, all it is doing is saying "what's the directory in this solution folder your other web project is".  Then it assumes the subfolder is wwwroot:
`<WRoot>$(SolutionDir)$(DestinationWebProjectName)\wwwroot\</WRoot>`

It shouldn't care whether or not the destination is a generic folder or an actual VS project, it's just that the way I am building the path currently assumed it's a project under the solution directory, so you can adjust WRoot to hardcode any destination path you want the WASM dist folder copied to on build.  There's nothing magical about DestinationWebProjectName that forces the destination to actually be a web project.  

 It also will look for a package_ folder in that destination and **delete it to cleanup old copies.  So use with caution/have a backup in case you mangle something and it deletes more than desired the first time.**

I don't use this approach locally anymore, but I would use it for test/release builds so it is all part of the same web app.  

When running locally, I use `<PackageReference Include="Uno.Wasm.Bootstrap.DevServer" Version="8.0.3" />` package and the C# project runs its own self hosted server, and from my other web project the script tag that reference embedded.js will use the full absolute URL of the C# WASM self host, i.e. `https://localhost:5000/embedded.js`.  So when running locally I have two self hosted apps running.   My main web application running, and then a second web host self hosting the WASM files.  The only reason I started doing it this way is to support breakpoints/debugging in the WASM.  I'm not sure how hard that would be to get working from the rails project because it depends on some JS for browserlink.  I'm sure if you figured out exactly what dependencies/JS needs to be loaded and how to inject the browserlink URL, then you'd be able to get the WASM breakpoints working to (see https://github.com/unoplatform/Uno.Wasm.Bootstrap/discussions/806).

**Reply by SerratedSharp** [reply]
Sorry, I just now noticed "output from my windows PC to my rails PC, and then to deploy it."  Unless you share a directory from the other PC then I'm not sure. 

You probably need to read up on how to host your own NPM package repository(yarn just uses NPM under the hood right?).  I'm a nuget guy so not enough of an expert in NPM to know if this is a good article or not, but something along these lines: https://medium.com/@debshish.pal/publish-a-npm-package-locally-for-testing-9a00015eb9fd

My msbuild script could be adjusted to copy to whatever shared destination where your repo would pack/prepare the package.  Then you could even add commands such as `npm pack` or whatever is needed to produce a npm package and register it.  I think you'd need some additional npm config files in the /dist/ to make it something that NPM recognizes as something that can be "packed".  I'd love to hear if you get it to work.  I think WASM/WASI is actually gonna be the future of NPM packages for a number of reasons, and publishing NPM packages is something I want to get to know better.

You might not actually need a NPM repo and just use shared folders.  It might be as simple as just putting the files in the right place under node modules or packages with the correct folder naming convention and a config file.  There's an old project I have that seemed to shared output from different projects to other projects by placing files in the appropriate packages folders, but there were enough layers of yarn scripts that it was beyond my abilities to decipher.

**Reply by benkoshy** [reply]
Hey I appreciate your posts and the links - all very relevant for study.

If I get this up and working, it will definitely warrant some documentation i will write this down somewhere or add a PR. Or perhaps there is some documentation out there, which I can approach given your pointers.

Much appreciate every single pointer you have given.

**Reply by SerratedSharp** [reply]
Yeh, I think if you view the wasm output as just static files, then any documentation out there for packing/sharing static files via NPM or maybe yarn workspaces is applicable.  I'm imagining there must be something that nodejs dev teams might do to share static resources between projects within their local network.

**Reply by benkoshy** [reply]
> Yeh, I think if you view the wasm output as just static files, then any documentation out there for packing/sharing static files via NPM or maybe yarn workspaces is applicable. I'm imagining there must be something that nodejs dev teams might do to share static resources between projects within their local network.

You were 100% on the money:  https://rustwasm.github.io/wasm-pack/book/tutorials/npm-browser-packages/packaging-and-publishing.html

Although I do not know the specifics, I think there is a way forward. I"ll make a PR if i'm successful at it.

**Reply by sasakrsmanovic** [reply]
Hey @benkoshy , did you ever get around this? Would love to close the discussion. Or maybe there was a PR about it that I didn't see. Thanks in either case!!

**Reply by benkoshy** [reply]
@sasakrsmanovic - I am very very appreciative of the guidance here, as will the rest of the community. 

You are correct in estimating that I was pulled from my c# implementation and the PR was not made......however the 'hello world' implementation will be crucial for me (and probably others in the rails community); your generous guidance alone DESERVES the courtesy of a hello world + guidance to others following. I will try and squeeze this out. please leave it with me.

**Reply by sasakrsmanovic** [reply]
Hey, no worries! I was just going through the list of open Q&A and noticed this one that can potentially be closed. It would be AMAZING to get that PR and doc contribution. Many on our team will be there to hep you i.e. @agneszitte or @MartinZikmund . Thanks in advance!!

**Reply by agneszitte** [reply]
> Hey, no worries! I was just going through the list of open Q&A and noticed this one that can potentially be closed. It would be AMAZING to get that PR and doc contribution. Many on our team will be there to hep you i.e. @agneszitte or @MartinZikmund . Thanks in advance!!

Hi @benkoshy, I was wondering if any additional help was needed here from @MartinZikmund or me. Please let us know :)

**Reply by benkoshy** [reply]
thanks for your msg. I will most certainly msg if there is an issue. if not me, someone in the rails community will face that problem.
